datashape/predicates.py

```diff
@@ -104,7 +109,6 @@ def _dimensions(ds):
         return 1 + _dimensions(ds.subshape[0])
     if isscalar(ds):
         return 0
-    raise NotImplementedError('Can not compute dimensions for %s' % ds)
```

Comment:
should this just return `None` if none of these checks catch? If `None` is valid, then we should have an explicit return here. If at least one of branches is expected, then we should replace this with an `AssertionError` so that it is easier to debug in the future.